### PR TITLE
Rewrite Conditional nodes before optimisation

### PR DESCRIPTION
**Pull request by tj-sun**
Handle Conditional nodes by the following rewriting:

  Conditional(condition, 1, 0) * THEN + Conditional(condition, 0, 1) * ELSE

this ensures Conditional nodes do not have argument indices, and
also factorisation can occur across THEN and ELSE branches

issue #119 

BTW I notice a `NAN` from this test
`test tests/extrusion/test_cylinder.py::test_betti2_cylinder[horiz_complex0-vert_complex0]`
C code:
`double  t0  = (cell_orientations[0][0] == 1 ? -1 : (cell_orientations[0][0] == 0 ? 1 : NAN));`
not sure is that expected?

### CODE REVIEW

**Comment by miklos1** [comment]
```c
double t0 = (cell_orientations[0][0] == 1 ? -1 : (cell_orientations[0][0] == 0 ? 1 : NAN));
```
This is fine, and unrelated to the current issue. `cell_orientations` expected to have value 0 or 1, which is translated to +1 and -1. If `cell_orientations` have unexpected value, we continue with a `NaN`.

**Comment by miklos1** [comment]
This is fine, however, can you decouple conditional factorisation from argument factorisation?

File: tsfc/coffee_mode.py [hunk 1]
```diff
@@ -33,13 +33,11 @@ def _handle_conditional(node, self):
...
 @_handle_conditional.register(Conditional)
 def _handle_conditional_conditional(node, self):
-    if not self.predicate(node):
-        return node
-    condition, then, else_ = node.children
-    then = self(then)
-    else_ = self(else_)
-    return Sum(Product(Conditional(condition, one, Zero()), then),
-               Product(Conditional(condition, Zero(), one), else_))
+    condition, then, else_ = map(self, node.children)
```
Comment (miklos1) [inline tsfc/coffee_mode.py]: Great use of `map`!

File: tsfc/coffee_mode.py [hunk 2]
```diff
@@ -33,13 +33,11 @@ def _handle_conditional(node, self):
...
+    if self.predicate(node):
+        return Sum(Product(Conditional(condition, one, Zero()), then),
+                   Product(Conditional(condition, Zero(), one), else_))
+    return Conditional(condition, then, else_)
```
Comment (miklos1) [inline tsfc/coffee_mode.py]: `reuse_if_untouched` better then "reconstruct always"

Comment (tj-sun) [inline tsfc/coffee_mode.py]: Ah I forgot. Amended.

**Comment by tj-sun** [comment]
Something like this?
```
    node = reuse_if_untouched(node, self)
    if self.predicate(node):
        condition, then, else_ = node.children
        return Sum(Product(Conditional(condition, one, Zero()), then),
                   Product(Conditional(condition, Zero(), one), else_))
    return node
```

File: tsfc/coffee_mode.py [hunk 3]
```diff
@@ -21,6 +23,39 @@
...
+    so that factorisation can occur across THEN and ELSE branches.
+    :arg expressions: list of GEM DAGs
+    :arg predicate: a predicate function on :class:`Conditional`s to determine
+                    whether to rewrite the node or not
+
+    :return: list of GEM DAGs with :class:`Conditional` nodes rewritten
+    """
+    mapper = Memoizer(_handle_conditional)
+    mapper.predicate = predicate
+    return list(map(mapper, expressions))
```
Comment (miklos1) [inline tsfc/coffee_mode.py]: Let's move this to `gem.optimise` under the name `factorise_conditional`.

File: tsfc/coffee_mode.py [hunk 4]
```diff
@@ -21,6 +23,39 @@
...
+def handle_conditional(expressions, predicate):
+    """Rewrite :class:`Conditional` nodes as:
+       Conditional(condition, 1, 0) * THEN + Conditional(condition, 0, 1) * ELSE
+    so that factorisation can occur across THEN and ELSE branches.
+    :arg expressions: list of GEM DAGs
+    :arg predicate: a predicate function on :class:`Conditional`s to determine
+                    whether to rewrite the node or not
+
+    :return: list of GEM DAGs with :class:`Conditional` nodes rewritten
+    """
```
Comment (miklos1) [inline tsfc/coffee_mode.py]: Docstring needs some stylistic and content fixes.

File: tsfc/coffee_mode.py [hunk 5]
```diff
@@ -45,15 +80,21 @@ def predicate(index):
     expressions = replace_delta(expressions)
     expressions = remove_componenttensors(expressions)
     expressions = replace_division(expressions)
-    return optimise_expressions(expressions, argument_multiindices)
+    argument_indices = tuple(itertools.chain(*argument_multiindices))
+
+    def conditional_predicate(argument_indices, node):
+        return len(set(node.free_indices) & argument_indices) >= 1
```
Comment (miklos1) [inline tsfc/coffee_mode.py]: If you can rewrite this predicate in terms of the classifier, then the `handle_conditional` call below may go to `collect_monomials`.

File: gem/refactorise.py [hunk 6]
```diff
@@ -229,6 +230,10 @@ def collect_monomials(expressions, classifier):
                                       predicate=lambda i: i in must_unroll)
         expressions = remove_componenttensors(expressions)
 
+    # Expand Conditional nodes which are COMPOUND
+    conditional_predicate = lambda node: classifier(node) == COMPOUND
```
Comment (miklos1) [inline gem/refactorise.py]: Does this pass `flake8`?

Comment (tj-sun) [inline gem/refactorise.py]: Yes... why?

Comment (miklos1) [inline gem/refactorise.py]: There is a rule against assigning `lambda`s to variables. But I guess we ignore that then.

Comment (tj-sun) [inline gem/refactorise.py]: I see. Good to know.

**Pull request merged**